core/src/cgroup.rs

```diff
@@ -1,238 +1,392 @@
+
+        cg1.mv(pid).unwrap();
+        let pids = cg1.get_pids().unwrap();
+        assert!(!pids.is_empty());
+        assert!(cg2.get_pids().unwrap().is_empty());
+        assert_eq!(pids.len(), 1);
+        assert_eq!(pids[0], pid);
+
+        cg2.mv(pid).unwrap();
+        let pids = cg2.get_pids().unwrap();
+        assert!(!pids.is_empty());
+        assert!(cg1.get_pids().unwrap().is_empty());
+        assert_eq!(pids.len(), 1);
+        assert_eq!(pids[0], pid);
+
+        proc.kill().unwrap();
+
+        cg1.rm().unwrap();
+    }
+
+    #[test]
+    fn populated() {
+        let mut proc = spawn_proc().unwrap();
+        let pid = Pid::from_raw(proc.id() as i32);
+        let cg = CGroup::new_root(Path::new(&mount_point().unwrap()), "cgroup_test").unwrap();
+
+        assert!(!cg.populated().unwrap());
+        assert_eq!(cg.populated().unwrap(), cg.get_pids().unwrap().len() > 0);
+
+        cg.mv(pid).unwrap();
+        assert!(cg.populated().unwrap());
+        assert_eq!(cg.populated().unwrap(), cg.get_pids().unwrap().len() > 0);
+
+        proc.kill().unwrap();
+
+        cg.rm().unwrap();
+    }
+
+    #[test]
+    fn frozen() {
+        let mut proc = spawn_proc().unwrap();
+        let pid = Pid::from_raw(proc.id() as i32);
+        let cg = CGroup::new_root(Path::new(&mount_point().unwrap()), "cgroup_test").unwrap();
+
+        // Freeze an empty cgroup
+        assert!(!cg.frozen().unwrap());
+        cg.freeze().unwrap();
+        assert!(cg.frozen().unwrap());
+
+        // Unfreeze the empty cgroup
+        cg.unfreeze().unwrap();
+        assert!(!cg.frozen().unwrap());
+
+        // Do the same with a non-empty cgroup
+        cg.mv(pid).unwrap();
+        cg.freeze().unwrap();
+        assert!(cg.frozen().unwrap());
+        cg.unfreeze().unwrap();
+        assert!(!cg.frozen().unwrap());
+
+        proc.kill().unwrap();
+
+        cg.rm().unwrap();
+    }
+
+    #[test]
+    fn kill() {
+        let proc = spawn_proc().unwrap();
+        let pid = Pid::from_raw(proc.id() as i32);
+        let cg = CGroup::new_root(Path::new(&mount_point().unwrap()), "cgroup_test").unwrap();
+
+        // Kill an empty cgroup
+        cg.kill().unwrap();
+
+        // Do the same with a non-empty cgroup
+        cg.mv(pid).unwrap();
+        assert!(cg.populated().unwrap());
+        cg.kill().unwrap();
+
+        cg.rm().unwrap();
+
+        // TODO: Check if the previous PID still exists (although unstable because the OS may re-assign)
+    }
+
+    #[test]
+    fn is_cgroup() {
+        assert!(super::is_cgroup(Path::new(&mount_point().unwrap())).unwrap());
+        assert!(!super::is_cgroup(Path::new("/tmp")).unwrap());
     }
-}
 
-impl From<PathBuf> for CGroup {
-    fn from(path: PathBuf) -> Self {
-        Self { path }
+    /// Spawns a child process of yes(1)
+    fn spawn_proc() -> io::Result<process::Child> {
+        process::Command::new("yes")
+            .stdout(process::Stdio::null())
+            .spawn()
```

Comment:
At least on my system, when running tests without `sudo`, some tests fail.
For some reason, the `yes` command spawned in these failed tests stays alive even after `cargo test` finishes.
Taking a look at the parent-pid(PPid) `cat /proc/{yes pid}/status | grep PPid` it says the parent process is pid 1.

Reply:
See 1d94d0f